Cerberus RMF Sensor Config 1

acschang opened the pull request:
Final modifications for the RMF Sensor Configuration 1 model submission by CERBERUS.

Original PR: #611

osrf-jenkins commented:
Build finished. 15 tests run, 0 skipped, 0 failed.

osrf-jenkins commented:
Build finished. 15 tests run, 0 skipped, 0 failed.

osrf-jenkins commented:
Build finished. No test results found.

osrf-jenkins commented:
Build finished. No test results found.

mjcarroll commented on submitted_models/cerberus_rmf_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,592 @@
+                    <horizontal_fov>2.8</horizontal_fov>
+                    <lens>
+                        <intrinsics>
+                            <fx>285.727813720703</fx>
```
These focal length intrinsics don't line up with our general formula:

`fx = fy = width / ( 2 * tan (hfov / 2 ) )`  which equals:  73.13

Is this intended, or from using actual camera calibration parameters?

mjcarroll replied:
CC: @mihirk284

mjcarroll replied:
I additionally went ahead and zeroed out the distortion parameters, because they aren't simulated in Ignition, and propagating them to ROS's camera_info could cause people to interpret the images as distorted.

mihirk284 replied:
@mjcarroll @nkoenig  thank you for iterating, and apologies for the delay.
The camera calibration parameters are taken from the actual sensor onboard the RMF robot.

I will quickly reverify them since they seem a bit off

acschang replied:
@mihirk284 Did you verify the camera calibration parameters match the ones used on the RMF robot?

mjcarroll commented on submitted_models/cerberus_rmf_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,592 @@
+                    <horizontal_fov>2.8</horizontal_fov>
+                    <lens>
+                        <intrinsics>
+                            <fx>285.194305419922</fx>
```
Same here

mjcarroll commented:
Verified sensors are available via ROS:

* Front left camera (+ optical frame)
* Front right camera (+ optical frame)
* Air pressure
* 4 lasers (back left, back right, front left, front right)
* IMU
* Gas sensor

Was able to fly via the `/rmf/cmd_vel` topic

mjcarroll reviewed (commented):
Overall, LGTM.  I have one question about the discrepancy between the camera parameters as listed versus what we would generally compute.

osrf-jenkins commented:
Build finished. No test results found.

osrf-jenkins commented:
Build finished. No test results found.

nkoenig commented:
@mihirk284, could you please address the above comments? Let me know if you have any question or need assistance.

mihirk284 commented:
@nkoenig @mjcarroll  The intrinsics of the sensor match the real sensor, however, I was not able to model the fisheye lens accurately for the submitted `model.sdf` file and hence had to manually change the horizontal field of view for the image to be understandable to the operator.

Please let me know what would be the appropriate way to correct this representation...

The camera calibration parameters of the real camera onboard the robot are as follows:-
```yaml
%YAML:1.0
# Input-output configuration
input: [848, 800]
output: [848, 800]

# Camera calibration parameter
K1: !!opencv-matrix
   rows: 3
   cols: 3
   dt: d
   data: [ 285.727813720703, 0.0, 414.698791503906,
           0.0, 285.767211914062, 405.671203613281, 
           0.0, 0.0, 1.0 ]
D1: !!opencv-matrix
   rows: 1
   cols: 4
   dt: d
   data: [ -0.00759695982560515, 0.0468390397727489, -0.0436895191669464, 0.00806778576225042 ]
K2: !!opencv-matrix
   rows: 3
   cols: 3
   dt: d
   data: [ 285.194305419922, 0.0, 415.857696533203,
           0.0, 285.091400146484, 403.501586914062, 
           0.0, 0.0, 1.0 ]
D2: !!opencv-matrix
   rows: 1
   cols: 4
   dt: d
   data: [-0.00481273513287306, 0.0396431498229504, -0.0368566811084747, 0.00591867882758379]
R: !!opencv-matrix
   rows: 3
   cols: 3
   dt: d
   data: [0.999975,-0.00321442, -0.00631769,
         0.00322045, 0.999994, 0.000945303,
         0.00631461, -0.000965625, 0.99998 ]
T: [ -0.0643780156970024, 0.000170402970979922, -0.000189287995453924 ]
```

mjcarroll commented:
I believe that in this case, since the simulator is not modelling the same lens as the physical sensor, it makes the most sense to set the camera `K` matrix to reflect what the simulator is actually producing using the formula that I posted above in https://github.com/osrf/subt/pull/722#discussion_r554170237.  This will ensure that the matrix in the `camera_info` topic is consistent with what is being simulated.  Additionally, since there is no distortion in the simulator, we can zero out those parameters as I did in 
3ff3594

nkoenig commented:
The bounding box for this model is
```
Min[-0.1476 -0.162577 0.021015] Max[0.148911 0.163575 0.172173]
```

mihirk284 commented:
@mjcarroll I intended to push this new intrinsics info, however, I do not seem to have access to push to this branch. Can you update it based on the discussion above? Updated (cx, cy as well)
The intrinsics for both the left and the right lenses are as follows:-

```xml
<intrinsics>
    <fx>73.1301317527</fx>
    <fy>73.1301317527</fy>
    <cx>424</cx>
    <cy>400</cy>
    <s>1</s>
</intrinsics>
```

Please let me know if any further change is to be made to the models.

Thanks :)

acschang commented:
> @mjcarroll I intended to push this new intrinsics info, however, I do not seem to have access to push to this branch. Can you update it based on the discussion above? Updated (cx, cy as well)
> The intrinsics for both the left and the right lenses are as follows:-
> 
> ```
> <intrinsics>
>     <fx>73.1301317527</fx>
>     <fy>73.1301317527</fy>
>     <cx>424</cx>
>     <cy>400</cy>
>     <s>1</s>
> </intrinsics>
> ```
> 
> Please let me know if any further change is to be made to the models.
> 
> Thanks :)

Thank you, I updated the parameters on the PR.  Please let us know if anything else needs to be updated.  The typical workflow for PRs once we create our own branch is to create another PR to merge into this PR's branch (i.e. PR into `submitted_models/cerberus_rmf_sensor_config_1` from your repository).

osrf-jenkins commented:
Build finished. 15 tests run, 0 skipped, 0 failed.
